homeassistant/components/greencell/config_flow.py

```diff
@@ -0,0 +1,119 @@
+"""Config flow for Greencell EVSE integration in Home Assistant.
+
+This module implements:
+- EVSEConfigFlow: an asyncio-based MQTT discovery flow that:
+  * Subscribes to the GREENCELL_DISC_TOPIC to listen for device response messages.
+  * Publishes a BROADCAST discovery request to GREENCELL_BROADCAST_TOPIC.
+  * Waits up to DISCOVERY_TIMEOUT seconds for a response, then aborts or continues.
+  * Extracts the device serial number from the discovery payload.
+  * Determines the device model name (Habu Den or Other) from the serial prefix.
+  * Registers a unique ID and creates the config entry with the serial_number.
+"""
+
+import asyncio
+from collections.abc import Callable
+import json
+import logging
+from typing import Any
+
+from greencell_client.utils import GreencellUtils
+
+from homeassistant import config_entries
+from homeassistant.components import mqtt
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import (
+    DISCOVERY_TIMEOUT,
+    DOMAIN,
+    GREENCELL_BROADCAST_TOPIC,
+    GREENCELL_DISC_TOPIC,
+    GREENCELL_HABU_DEN,
+    GREENCELL_OTHER_DEVICE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class EVSEConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Config flow for EVSE device."""
+
+    _remove_listener: Callable[[], None] | None
+    VERSION = 1
+    MINOR_VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self.discovery_event = asyncio.Event()
+        self.discovery_data = None
+        self._remove_listener: Callable[[], None] | None = None
+
+    def get_device_name(self, serial_number: str) -> str:
+        """Get device name based on serial number."""
+        if GreencellUtils.device_is_habu_den(serial_number):
+            return GREENCELL_HABU_DEN
+        return GREENCELL_OTHER_DEVICE
+
+    async def _publish_disc_request(self):
+        """Publish a discovery request to the MQTT topic."""
+        payload = json.dumps({"name": "BROADCAST"})
+        await mqtt.async_publish(self.hass, GREENCELL_BROADCAST_TOPIC, payload, 0, True)
+
+    @callback
+    def _mqtt_message_received(self, msg):
+        """Handle incoming MQTT messages."""
+        try:
+            payload = json.loads(msg.payload)
+            self.discovery_data = payload
+            self.discovery_event.set()
+        except json.JSONDecodeError:
+            _LOGGER.error("Failed to decode MQTT message payload: %s", msg.payload)
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> config_entries.ConfigFlowResult:
+        """Handle the initial step."""
```

Comment:
You should first check that mqtt is configured and connected to the MQTT broker, that can be done by calling mqtt.is_connected as done in the [pglab integration](https://github.com/home-assistant/core/blob/25806615a944206be9f339c873cf9a99011a6ffc/homeassistant/components/pglab/config_flow.py#L41-L42)